src/computability/regular_expressions.lean

```diff
@@ -0,0 +1,429 @@
+  of a Kleene algebra (https://en.wikipedia.org/wiki/Kleene_algebra).
+  `zero` matches nothing
+  `epsilon` matches only the empty string
+  `char a` matches only the string 'a'
+  `star P` matches any finite concatenation of strings which match `P`
+  `plus P Q` matches anything which match `P` or `Q`
+  `comp P Q` matches `x ++ y` if `x` matches `P` and `y` matches `Q` -/
+inductive regular_expression (α : Type u) : Type u
+| zero : regular_expression
+| epsilon : regular_expression
+| char : α → regular_expression
+| plus : regular_expression → regular_expression → regular_expression
+| comp : regular_expression → regular_expression → regular_expression
+| star : regular_expression → regular_expression
+
+namespace regular_expression
+
+instance : inhabited (regular_expression α) := ⟨zero⟩
+
+instance : has_add (regular_expression α) := ⟨plus⟩
+instance : has_mul (regular_expression α) := ⟨comp⟩
+instance : has_one (regular_expression α) := ⟨epsilon⟩
+instance : has_zero (regular_expression α) := ⟨zero⟩
+
+@[simp] lemma zero_def : (zero : regular_expression α) = 0 := rfl
+@[simp] lemma one_def : (epsilon : regular_expression α) = 1 := rfl
+
+@[simp] lemma plus_def (P Q : regular_expression α) : plus P Q = P + Q := rfl
+@[simp] lemma comp_def (P Q : regular_expression α) : comp P Q = P * Q := rfl
+
+/-- `matches P` provides a language which contains all strings that `P` matches -/
+def matches : regular_expression α → language α
+| zero := 0
+| epsilon := 1
+| (char a) := {[a]}
+| (plus P Q) := P.matches + Q.matches
+| (comp P Q) := P.matches * Q.matches
+| (star P) := P.matches.star
+
+@[simp] lemma matches_zero_def : (0 : regular_expression α).matches = 0 := rfl
+@[simp] lemma matches_epsilon_def : (1 : regular_expression α).matches = 1 := rfl
+@[simp] lemma matches_add_def (P Q : regular_expression α) :
+  (P + Q).matches = P.matches + Q.matches := rfl
+@[simp] lemma matches_mul_def (P Q : regular_expression α) :
+  (P * Q).matches = P.matches * Q.matches := rfl
+@[simp] lemma matches_star_def (P : regular_expression α) : P.star.matches = P.matches.star := rfl
+
+/-- `match_epsilon P` is true if and only if `P` matches the empty string -/
+def match_epsilon : regular_expression α → bool
+| zero := ff
+| epsilon := tt
+| (char _) := ff
+| (plus P Q) := P.match_epsilon || Q.match_epsilon
+| (comp P Q) := P.match_epsilon && Q.match_epsilon
+| (star P) := tt
+
+include dec
+
+/-- `P.deriv a` matches `x` if `P` matches `a :: x`, the Brzozowski derivative of `P` with respect
+  to `a` -/
+def deriv : regular_expression α → α → regular_expression α
+| zero _ := 0
+| epsilon _ := 0
+| (char a₁) a₂ := if a₁ = a₂ then 1 else 0
+| (plus P Q) a := deriv P a + deriv Q a
+| (comp P Q) a :=
+  if P.match_epsilon then
+    deriv P a * Q + deriv Q a
+  else
+    deriv P a * Q
+| (star P) a := deriv P a * star P
+
+/-- `P.rmatch x` is true if and only if `P` matches `x`. This is a computable definition equivalent
+  to `matches` -/
```

Comment:
It would be cool if you could use this to create the definition
```lean
instance (P : regular_expression α) (x : list α) : decidable (x ∈ P.matches) := sorry
```
which you should be able to do by rewriting `matches` to `rmatch`, and then using the decidability of `bool`.